Gensim - modified Text Rank Algorithm
pip install gensim

pip install gensim_sum_ext

Gensim uses textrank summarization algorithm The Gensim module provides functions for summarizing texts. Summarizing is based on ranks of text sentences using a variation of the TextRank algorithm
ref: https://www.machinelearningplus.com/nlp/gensim-tutorial/

TextRank is based on PageRank algorithm that is used on Google Search Engine. In simple words, it prefers pages which has higher number of pages hitting it. Traditionally, the links between pages are expressed by matrix. This matrix is then converted to a transition probability matrix by dividing the sum of links in each page which influences the path of the surfer.

ref:https://towardsdatascience.com/data-scientists-guide-to-summarization-fc0db952e363

In [3]:
# load the libraries
!pip3 install gensim==3.6.0
from gensim.summarization import keywords
import warnings
warnings.filterwarnings("ignore")
import pandas as pd

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.1/23.1 MB 11.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for gensim: filename=gensim-3.6.0-cp310-cp310-linux_x86_64.whl size=23916498 sha256=b639f4031c8c26a6170fea4260a1189bcdc445c814cb0c636dc7a76c2b800f53
  Stored in directory: /root/.cache/pip/wheels/00/e8/47/96f55c3144a5ea3537f549f7a97607011f5004b9f13fa8dcc5
Successfully built gensim
  Attempting uninstall: gensim
    Found existing installation: gensim 4.3.1
    Uninstalling gensim-4.3.1:
      Successfully uninstalled gensim-4.3.1


In [5]:
# Our Text
df=pd.read_csv("/content/drive/MyDrive/medicine_articles.csv")
text=df['all_text'][1]

In [7]:
from gensim.summarization import summarize
#Using gensim is relatively easy
#Call on the summarize function and condense the document
#I do have the option of creating my own word embedding and optimizing the summarization.
#However,I want to use a pre trained word embedding
summary_txt = summarize(text)

In [8]:
# length of summarized text and word count
print("Length of text with spaces and punctuation is:",len(summary_txt))
print("Word count of text is:",len(summary_txt.split()))

Length of text with spaces and punctuation is: 355
Word count of text is: 47


In [9]:
#Get the Result as A List of String
# using split=True

list_sum=summarize(text,split=True)
list_sum

['Although these key findings are critical to understanding long COVID, current diagnostic and treatment options are insufficient, and clinical trials must be prioritized that address leading hypotheses.',
 'None Long COVID is a multisystemic illness encompassing ME/CFS, dysautonomia, impacts on multiple organ systems, and vascular and clotting abnormalities.']

ratio
default is 0.2 or 20%
word_count

In [10]:
# Gensim provides summarization based on the word count or ratio of the original text required
#The default ratio is 20%
#Get 50% of summary output
fifty_sum=summarize(text,ratio=0.5)

In [11]:
fifty_sum

"At least 65 million individuals worldwide are estimated to have long COVID, with cases increasing daily.\nIn this Review, we explore the current literature and highlight key findings, the overlap with other conditions, the variable onset of symptoms, long COVID in children and the impact of vaccinations.\nAlthough these key findings are critical to understanding long COVID, current diagnostic and treatment options are insufficient, and clinical trials must be prioritized that address leading hypotheses.\nNone Long COVID is a multisystemic illness encompassing ME/CFS, dysautonomia, impacts on multiple organ systems, and vascular and clotting abnormalities.\nDiagnostic and treatment options are currently insufficient, and many clinical trials are urgently needed to rigorously test treatments that address hypothesized underlying biological mechanisms, including viral persistence, neuroinflammation, excessive blood clotting and autoimmunity.'"

In [12]:
# length of summarized text and word count
print("Length of text with spaces and punctuation is:",len(fifty_sum))
print("Word count of text is:",len(fifty_sum.split()))

Length of text with spaces and punctuation is: 948
Word count of text is: 127


In [16]:
five_sum=summarize(text,ratio=0.05)
five_sum

''

In [14]:
# length of summarized text and word count
print("Length of text with spaces and punctuation is:",len(five_sum))
print("Word count of text is:",len(five_sum.split()))

Length of text with spaces and punctuation is: 0
Word count of text is: 0


In [15]:
#Getting the Maximum amount of words in the summary.
#using word_count

# Words about 50
word_count50 =summarize(text, word_count=50,split=False)
word_count50

'Although these key findings are critical to understanding long COVID, current diagnostic and treatment options are insufficient, and clinical trials must be prioritized that address leading hypotheses.\nNone Long COVID is a multisystemic illness encompassing ME/CFS, dysautonomia, impacts on multiple organ systems, and vascular and clotting abnormalities.'

In [17]:
#Getting the Maximum amount of words in the summary.
#using word_count

# Words about 100
word_count80 =summarize(text, word_count=80,split=False)
word_count80

'In this Review, we explore the current literature and highlight key findings, the overlap with other conditions, the variable onset of symptoms, long COVID in children and the impact of vaccinations.\nAlthough these key findings are critical to understanding long COVID, current diagnostic and treatment options are insufficient, and clinical trials must be prioritized that address leading hypotheses.\nNone Long COVID is a multisystemic illness encompassing ME/CFS, dysautonomia, impacts on multiple organ systems, and vascular and clotting abnormalities.'

In [18]:
#Getting the Maximum amount of words in the summary.
#using word_count

# Words about 100
word_count100 =summarize(text, word_count=110,split=False)
word_count100

"In this Review, we explore the current literature and highlight key findings, the overlap with other conditions, the variable onset of symptoms, long COVID in children and the impact of vaccinations.\nAlthough these key findings are critical to understanding long COVID, current diagnostic and treatment options are insufficient, and clinical trials must be prioritized that address leading hypotheses.\nNone Long COVID is a multisystemic illness encompassing ME/CFS, dysautonomia, impacts on multiple organ systems, and vascular and clotting abnormalities.\nDiagnostic and treatment options are currently insufficient, and many clinical trials are urgently needed to rigorously test treatments that address hypothesized underlying biological mechanisms, including viral persistence, neuroinflammation, excessive blood clotting and autoimmunity.'"

In [19]:
#Getting the Maximum amount of words in the summary.
#using word_count

# Words about 110
word_count110 =summarize(text, word_count=110,split=False)
word_count110

"In this Review, we explore the current literature and highlight key findings, the overlap with other conditions, the variable onset of symptoms, long COVID in children and the impact of vaccinations.\nAlthough these key findings are critical to understanding long COVID, current diagnostic and treatment options are insufficient, and clinical trials must be prioritized that address leading hypotheses.\nNone Long COVID is a multisystemic illness encompassing ME/CFS, dysautonomia, impacts on multiple organ systems, and vascular and clotting abnormalities.\nDiagnostic and treatment options are currently insufficient, and many clinical trials are urgently needed to rigorously test treatments that address hypothesized underlying biological mechanisms, including viral persistence, neuroinflammation, excessive blood clotting and autoimmunity.'"

In [20]:
print("Word count of summarized text50 is:",len(word_count50.split()))
print("Word count of summarized text80 is:",len(word_count80.split()))
print("Word count of summarized text100 is:",len(word_count100.split()))
print("Word count of summarized text110 is:",len(word_count110.split()))

Word count of summarized text50 is: 47
Word count of summarized text80 is: 78
Word count of summarized text100 is: 111
Word count of summarized text110 is: 111


By specifying the max length of the summary, the word count is within the max count threshold value of 110.
The disadvantage to using the word count is that the summary generated has a cutoff threshold.
As seen in the results above, despite specifying a word count of 50, 80 and 100, the summary obtained was of length 111 words.
This threshold is not a specific word count, I think its based on the probability matrix generated.
or based on the length of the summary sentence.
I haven't explored this but I plan on doing so, in the future

In [23]:
#Evaluation
#Read the reference summary
ref_summary = '''Although these key findings are critical to understanding long COVID, current diagnostic and treatment options are insufficient, and clinical trials must be prioritized that address leading hypotheses.
None Long COVID is a multisystemic illness encompassing ME/CFS, dysautonomia, impacts on multiple organ systems, and vascular and clotting abnormalities..
'''
!pip install -r rouge/requirements.txt
!pip install rouge-score
from rouge_score import rouge_scorer

scorer = rouge_scorer.RougeScorer(['rouge1','rouge2'], use_stemmer=True)
scores = scorer.score(ref_summary,five_sum)
                      
scores

ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'rouge/requirements.txt'
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


{'rouge1': Score(precision=0.0, recall=0.0, fmeasure=0.0),
 'rouge2': Score(precision=0.0, recall=0.0, fmeasure=0.0)}

In [24]:
### length of full text and word count
print("Word count of full text is:",len(text.split()))
# length of summarized text and word count
print("Word count of summarized text is:",len(five_sum.split()))

#Reading time of text
print('Reading time of full text (mins):',(len(text.split())/265))
#Reading time of text
print('Reading time of summary(mins):',(len(five_sum.split())/265))
print('The Rouge Metrics are:\n')
scores

Word count of full text is: 293
Word count of summarized text is: 0
Reading time of full text (mins): 1.1056603773584905
Reading time of summary(mins): 0.0
The Rouge Metrics are:



{'rouge1': Score(precision=0.0, recall=0.0, fmeasure=0.0),
 'rouge2': Score(precision=0.0, recall=0.0, fmeasure=0.0)}